README.rst
A file handlers that read and write encryptes files.


.. image:: https://secure.travis-ci.org/efagerberg/crypto-file.png
    :target: https://secure.travis-ci.org/efagerberg/crypto-file/

.. image:: https://coveralls.io/repos/github/efagerberg/crypto-file/badge.svg?branch=master
    :target: https://coveralls.io/github/efagerberg/crypto-file?branch=master



Requirements
============

- Python 2.7

Usage
=====

Reading
-------

.. code-block:: python

    from crypto_file import Reader
    with Reader(fname='file.txt', password='Foo') as f:
        f.read()

Writing
-------

.. code-block:: python

    from crypto_file import Writer
    with Writer(fname='file.txt', password='Foo') as f:
        f.write("Some secret stuff")

CHANGELOG
Version 1.0.0
 * 100% test coverage

crypto_file/tests/test_reader.py
import mock
import pytest

from crypto_file.reader import Reader


@pytest.fixture()
def mock_open():
    with mock.patch('__builtin__.open') as m:
        yield m


@pytest.fixture()
def reader(mock_open):
    mock_open.return_value.read.return_value = 'Salted__asdfghj\n'
    reader = Reader(fname='foo.txt',
                    password='foo')
    mock_open.reset_mock()
    yield reader


def test_init_errors_without_password_or_key():
    with pytest.raises(ValueError):
        Reader(fname='foo')


def test_init_generates_expected_attributes(reader):
    assert reader.salt is not None
    assert reader.iv is not None
    assert reader.cipher is not None
    assert reader.streamLines == 0
    assert reader.chunk_unprocessed == ''
    assert reader.enableSeek is True
    assert reader.prev_stream == ''


def test_get_salt_makes_salt_from_fObj(reader):
    expected_salt = 'asdfghj\n'
    reader = Reader(fname='foo.txt',
                    password='foo')
    reader.get_salt()

    assert reader.salt == expected_salt


def test_readline_returns_chunk_when_file_is_closed(reader):
    reader.fileOpen = False
    reader.stream = "FooBar\nBaz"
    chunk = reader.readline() 

    assert chunk == 'FooBar\n'
    assert reader.streamLines == -1
    assert reader.prev_stream == chunk


def test_readline_decrypts_returns_chunk(reader, mock_open):
    reader.chunk_unprocessed = 'Something'
    reader.fObj.read.return_value = ''
    reader.stream = "FooBar\nBaz"
    chunk = reader.readline()

    assert chunk == 'FooBar\n'
    assert reader.streamLines == -1
    assert reader.prev_stream == chunk


def test_readline_does_not_set_prev_stream(reader, mock_open):
    reader.chunk_unprocessed = 'Something'
    reader.fObj.read.return_value = ''
    reader.stream = "FooBar\nBaz"
    reader.enableSeek = False
    chunk = reader.readline()

    assert chunk == 'FooBar\n'
    assert reader.streamLines == -1
    assert reader.prev_stream == ''


def test_read_returns_chunk_when_file_is_closed(reader):
    reader.fileOpen = False
    reader.stream = "FooBar\nBaz"
    chunk = reader.read(2)

    assert chunk == "Fo"
    assert reader.streamLines == 0
    assert reader.prev_stream == chunk


def test_read_decrypts_returns_chunk(reader, mock_open):
    reader.stream = "FooBar\nBaz"
    chunk = reader.read(2)

    assert chunk == 'Fo'
    assert reader.streamLines == 0
    assert reader.prev_stream == chunk


def test_read_does_not_set_prev_stream(reader, mock_open):
    reader.stream = "FooBar\nBaz"
    reader.enableSeek = False
    chunk = reader.read(2)

    assert chunk == 'Fo'
    assert reader.streamLines == 0
    assert reader.prev_stream == ''


def test_read_decrypts_chunk_when_size_larger_than_stream(reader, mock_open):
    expected_chunk = "FooBar\nBaz"
    reader.stream = expected_chunk
    chunk = reader.read(len(reader.stream) + 1)

    assert chunk == expected_chunk + '\x03'
    assert reader.streamLines == 0
    assert reader.prev_stream == expected_chunk + '\x03'


def test_read_with_no_size_gets_whole_stream(reader, mock_open):
    reader.chunk_unprocessed = ''
    expected_chunk = 'FooBar\nBaz\x03\x85\xd3\xf4'
    reader.fObj.read.side_effect = ('Salted__asdfghj\n', '')
    reader.stream = "FooBar\nBaz"
    chunk = reader.read()

    assert chunk == expected_chunk
    assert reader.streamLines == 0
    assert reader.prev_stream == expected_chunk


def test_seek_raises_error_when_enableSeek_False(reader):
    reader.enableSeek = False
    with pytest.raises(NotImplementedError):
        reader.seek(1)


def test_seek_eof_raises_error(reader):
    with pytest.raises(IOError):
        reader.seek(1, mode=2)


def test_seek_with_unknown_mode_raises_error(reader):
    with pytest.raises(IOError):
        reader.seek(1, mode=34)


def test_seek_with_mode_zero_reads_position_in_stream(reader):
    reader.prev_stream = "Foo"
    reader.seek(1)

    assert reader.stream == 'oo'
    assert reader.prev_stream == 'F'


def test_seek_with_mode_one_reads_stream(reader):
    reader.prev_stream = "Foo"
    reader.seek(1, mode=1)

    assert reader.stream != ''
    assert reader.prev_stream == 'Foo\x03'


def test_decrypt_chunk_appends_steam_one_line(reader):
    expected_stream = 'Bar'
    reader.chunk_unprocessed = expected_stream
    reader.fObj.read.return_value = '1' * 16
    reader.decrypt_chunk()

    assert reader.stream == expected_stream
    assert reader.streamLines == 0
    assert reader.fileOpen


def test_decrypt_chunk_appends_steam_two_line(reader):
    expected_stream = 'Bar\nBaz'
    reader.chunk_unprocessed = expected_stream
    reader.fObj.read.return_value = '1' * 16
    reader.decrypt_chunk()

    assert reader.stream == expected_stream
    assert reader.streamLines == 1
    assert reader.fileOpen


def test_decrypt_chunk_processes_chunk_when_no_cipher_left(reader):
    expected_stream = '1' * 128
    reader.chunk_unprocessed = expected_stream
    reader.fObj.read.return_value = ''
    reader.decrypt_chunk()

    assert reader.stream == expected_stream[:-49]
    assert reader.streamLines == 0
    assert not reader.fileOpen


crypto_file/tests/test_writer.py
import base64

import mock
import pytest

from crypto_file.writer import Writer


@pytest.fixture()
def mock_open():
    with mock.patch('__builtin__.open') as m:
        yield m


@pytest.fixture()
def writer(mock_open):
    writer = Writer(fname='foo.txt',
                    password='foo')
    mock_open.reset_mock()
    yield writer


def test_init_generates_expected_attributes(writer, mock_open):
    assert writer.salt is not None
    assert writer.iv is not None
    assert writer.cipher is not None


def test_init_creates_savekey_file(writer, mock_open):
    expected_file = 'foo.key'
    writer.__init__(fname='foo.txt', password='foo',
                    saveKey_file=expected_file)

    last_open_call = mock_open.call_args_list[-1]
    assert last_open_call == mock.call(expected_file, 'wb')


def test_gen_key_generates_key_w_no_initial_key_and_pass(writer, mock_open):
    writer.key = None
    writer.password = None
    writer.gen_key()

    assert writer.key is not None
    assert writer.password == base64.b64encode(writer.key)


def test_gen_key_generates_key_without_initial_key_only(writer, mock_open):
    writer.key = None
    writer.gen_key()

    assert writer.key is not None
    assert writer.password == base64.b64encode(writer.key)


def test_get_salt_generates_salt(writer, mock_open):
    writer.salt = None
    writer.get_salt()

    assert writer.salt is not None


def test_write_adds_input_to_steam_and_checks_buffer(writer):
    path = 'crypto_file.writer.Writer.check_write_buffer'
    with mock.patch(path) as m:
        writer.write("Foo")

    m.assert_called_once()
    assert writer.stream == "Foo"


def test_check_write_buffer_with_large_enough_stream(writer):
    writer.fObj = mock.Mock(spec=file)
    writer.stream = '1' * 1024 * writer.bs
    writer.check_write_buffer()

    writer.fObj.write.assert_called_once()
    assert writer.stream == ''


def test_close_does_nothing_when_file_already_closed(writer):
    writer.fObj = mock.Mock(spec=file)
    writer.fObj.closed = True
    writer.close()

    assert not writer.fObj.write.called
    assert not writer.stream


def test_close_handles_partial_stream(writer):
    writer.fObj = mock.Mock(spec=file)
    writer.fObj.closed = False
    # A partial stream
    writer.stream = '1' * (writer.bs + 1)
    writer.close()

    writer.fObj.write.assert_called_once()
    writer.fObj.close.assert_called_once()
    assert writer.stream != writer.bs * chr(writer.bs)


def test_close_with_no_partial_stream_writes_to_file(writer):
    writer.fObj = mock.Mock(spec=file)
    writer.fObj.closed = False
    writer.close()

    writer.fObj.write.assert_called_once()
    writer.fObj.close.assert_called_once()
    assert writer.stream == writer.bs * chr(writer.bs)


crypto_file/tests/test_crypto_handler.py
import base64

import mock
import pytest

from Crypto.Cipher import AES

from crypto_file.crypto_handler import CryptoHandler


BASE_MOCK_PATH = 'crypto_file.crypto_handler'


@pytest.fixture()
def mock_open():
    with mock.patch('{}.open'.format(BASE_MOCK_PATH)) as m:
        yield m


@pytest.fixture()
def handler(mock_open):
    handler = CryptoHandler(fname='foo.txt',
                            mode='rb',
                            password='foo')
    mock_open.reset_mock()
    yield handler


@pytest.fixture()
def mock_b64():
    with mock.patch('{}.base64'.format(BASE_MOCK_PATH), autospec=True) as m:
        m.b64decode.return_value = '1' * 32
        yield m


@pytest.fixture()
def mock_sha256():
    path = '{}.hashlib.sha256'.format(BASE_MOCK_PATH)
    with mock.patch(path, autospec=True) as m:
        yield m


def asset_good_handler(handler, mock_open, expected_fname, expected_mode):
    assert handler.mode == expected_mode
    assert handler.fileOpen is True
    assert handler.stream == ''
    assert handler.key is not None
    assert handler.password == base64.b64encode(handler.key)

    if isinstance(expected_fname, str):
        mock_open.assert_called_once_with(expected_fname, expected_mode)


def test_handler_constructor_handles_string_fname(mock_open):
    expected_fname = 'test.csv'
    mock_open.return_value.mode = expected_mode = 'r'
    handler = CryptoHandler(fname=expected_fname,
                            mode=expected_mode,
                            password='foo')

    asset_good_handler(handler, mock_open, expected_fname, expected_mode)


def test_handler_constructor_handles_file_fname_with_b(mock_open):
    mock_file = mock.Mock(spec=file)
    mock_file.mode = expected_mode = 'rb'
    handler = CryptoHandler(fname=mock_file,
                            mode=expected_mode,
                            password='foo')

    asset_good_handler(handler, mock_open, mock_file, expected_mode)


def test_handler_constructor_handles_file_fname_without_b(mock_open):
    mock_file = mock.Mock(spec=file)
    mock_file.mode = 'r'
    expected_mode = 'rb'
    mock_open.return_value.mode = expected_mode
    handler = CryptoHandler(fname=mock_file,
                            mode=expected_mode,
                            password='foo')

    asset_good_handler(handler, mock_open, mock_file, expected_mode)


def test_raises_io_error_without_correct_fname():
    with pytest.raises(IOError):
        CryptoHandler(fname=1)


def test_gen_key_handles_key_as_keyfile_sets_pw(handler, mock_open, mock_b64):
    handler.key = 'keyfile.key'
    handler.gen_key()

    assert handler.key == mock_open.return_value.read.return_value
    assert handler.password == mock_b64.b64encode.return_value
    mock_b64.b64encode.assert_called_once_with(handler.key)
    mock_open.assert_called_once_with('keyfile.key', 'rb')


def test_gen_key_handles_key_str(handler, mock_open, mock_b64):
    handler.key = 'foo'
    handler.gen_key()

    assert handler.key == mock_b64.b64decode.return_value
    assert handler.password == mock_b64.b64encode.return_value
    mock_b64.b64encode.assert_called_once_with(handler.key)
    assert not mock_open.called


def test_gen_key_errors_with_non_32_len_key(handler, mock_open, mock_b64):
    handler.key = '1'
    mock_b64.b64decode.return_value = 'NotLongEnough'
    with pytest.raises(NotImplementedError):
        handler.gen_key()

    assert not mock_b64.b64encode.called
    assert not mock_open.called


def test_gen_key_generates_key_from_password(handler, mock_b64):
    handler.gen_key()

    mock_b64.b64encode.assert_called_once_with(handler.key)
    assert handler.password == mock_b64.b64encode.return_value


def test_gen_key_errors_with_invalid_password(handler, mock_b64):
    handler.password = 1234
    handler.key = None
    with pytest.raises(ValueError):
        handler.gen_key()


def test_gen_iv_sets_iv(handler, mock_sha256):
    handler.salt = 'salty'
    mock_sha256.return_value.digest.return_value = b'1'
    handler.gen_iv()

    assert handler.iv == b'1' * AES.block_size


def test_check_mode_passes_if_mode_matches(handler):
    handler.mode = 'rb'

    handler.check_mode('r')
    handler.check_mode('rb')


def test_check_mode_errors_if_mode_does_not_match(handler):
    handler.mode = 'wb'

    with pytest.raises(IOError):
        handler.check_mode('r')


def test_close_closes_file(handler):
    handler.fObj = mock.Mock(spec=file)
    handler.close()

    assert handler.fObj.closed


def test__enter__returns_handler(handler):
    assert handler.__enter__() == handler


def test__exit__closes_file(handler):
    handler.fObj = mock.Mock(spec=file)
    handler.__exit__('foo', 'bar', 'traceback')

    assert handler.fObj.closed


def test__del__closes_file(handler):
    handler.fObj = fObj = mock.Mock(spec=file)
    del handler

    assert fObj.closed


crypto_file/reader.py
from Crypto.Cipher import AES

from crypto_file import CryptoHandler


class Reader(CryptoHandler):

    def __init__(self, fname, password=None, key=None, enableSeek=True):

        # Check for a key or password
        if key is None and password is None:
            msg = 'Need either a password or a key (file) for decryption'
            raise ValueError(msg)

        super(Reader, self).__init__(fname, password, key, 'rb')

        # Verify the file object is in read mode
        self.check_mode('r')

        # Need to retrieve the salt from the file
        self.get_salt()
        self.gen_iv()

        # Initiate the decryptor
        self.cipher = AES.new(self.key, AES.MODE_CBC, self.iv)

        # Setup the object's stream [used as a read buffer or write buffer]
        self.streamLines = 0
        self.chunk_unprocessed = ''
        self.enableSeek = enableSeek

        # Allow for seek function; need to store previous values
        self.prev_stream = ''

    # Retrieve or Create the salt for the cipher
    def get_salt(self):
        self.salt = self.fObj.read(self.bs)[len('Salted__'):]

    # Method to read the next line
    # + decrypt enough to cover the next line
    def readline(self):
        while self.streamLines == 0 and self.fileOpen:
            self.decrypt_chunk()

        currLine, lineSuff, self.stream = self.stream.partition('\n')
        currChunk = currLine + lineSuff
        self.streamLines -= 1

        if self.enableSeek:
            self.prev_stream += currChunk

        return currChunk

    def read(self, size=None):
        while (len(self.stream) < size or size is None) and self.fileOpen:
            self.decrypt_chunk()
        if size is None:
            size = len(self.stream)
            currChunk, self.stream = self.stream, ''
        else:
            size = min(size, len(self.stream))
            currChunk = self.stream[:size]
            self.stream = self.stream[size:]

        if self.enableSeek:
            self.prev_stream += currChunk

        return currChunk

    def seek(self, pos, mode=0):
        if not self.enableSeek:
            raise NotImplementedError('seek disabled for this object')

        # - Absolute position: from the start of the file
        # + equivalent to a relative position plus the
        #  length of the previous stream
        if mode == 0:
            pos -= len(self.prev_stream)

        # - Relative position: from the current position
        if mode == 0 or mode == 1:
            if pos < 0:
                self.stream = self.prev_stream[pos:] + self.stream
                self.prev_stream = self.prev_stream[:pos]
            else:
                self.read(pos)

        elif mode == 2:
            raise IOError('Seek from EOF is not implemented')
        else:
            raise IOError('Unknown mode requested for seek')

    def decrypt_chunk(self):
        self.chunk_processed = self.chunk_unprocessed
        crypted = self.fObj.read(1024 * self.bs)
        self.chunk_unprocessed = self.cipher.decrypt(crypted)
        if len(self.chunk_unprocessed) == 0:
            padding_length = ord(self.chunk_processed[-1])
            self.chunk_processed = self.chunk_processed[:-padding_length]
            self.fileOpen = False
        self.stream += self.chunk_processed
        self.streamLines += self.chunk_processed.count('\n')


crypto_file/writer.py
import hashlib
import base64

from Crypto.Cipher import AES
from Crypto import Random

from crypto_file import CryptoHandler


class Writer(CryptoHandler):

    def __init__(self, fname, password=None, key=None, saveKey_file=None):
        super(Writer, self).__init__(fname, password, key, 'wb')

        # Write the key to the specified file
        if saveKey_file is not None:
            open(saveKey_file, 'wb').write(self.key)

        # Verify the file object is in write mode
        self.check_mode('w')

        self.get_salt()
        self.gen_iv()

        # Initiate the encryptor
        self.cipher = AES.new(self.key, AES.MODE_CBC, self.iv)

    def gen_key(self):
        should_create_new_key = (self.key is None and
                                 self.password is None)
        if should_create_new_key:
            self.key = hashlib.sha256(Random.new().read(32)).digest()
            print('Encryption Key: {}\n'.format(base64.b64encode(self.key)))
            # Have to create a new password from the key to use in IV
            # - allows for message recovery using key file without password
            self.password = base64.b64encode(self.key)
        else:
            super(Writer, self).gen_key()

    # Create the salt for the cipher
    def get_salt(self):
        # If it is a writable file need to
        # save the salt value to the start of the file
        self.salt = Random.new().read(self.bs - len('Salted__'))
        self.fObj.write('Salted__' + self.salt)

    def write(self, s):
        self.stream += s
        self.check_write_buffer()

    # Check if enough data is available to write an encrypted chunk
    def check_write_buffer(self):
        # - retain partial chunks and only pad if the file is being closed
        while len(self.stream) >= 1024 * self.bs:
            chunk = self.stream[:1024*self.bs]
            self.fObj.write(self.cipher.encrypt(chunk))
            self.stream = self.stream[1024*self.bs:]

    def close(self):
        if self.fObj.closed:
            return

        # Check for a partial string and pad if necessary
        if len(self.stream) == 0 or len(self.stream) % self.bs != 0:
            padding_length = (self.bs - len(self.stream) % self.bs) or self.bs
            self.stream += padding_length * chr(padding_length)
        else:
            self.fObj.write(self.cipher.encrypt(self.stream))
            self.stream = self.bs * chr(self.bs)

        self.fObj.write(self.cipher.encrypt(self.stream))
        super(Writer, self).close()


crypto_file/crypto_handler.py
import hashlib
import base64

from Crypto.Cipher import AES

#  AES encrypted file-like object
#  - Support for Reader and Writer objects


class CryptoHandler(object):
    def __init__(self, fname, password=None, key=None, mode=None):

        # Create a file handler
        if isinstance(fname, str):
            self.fObj = open(fname, mode)
        elif isinstance(fname, file):  # noqa: F821
            curr_mode = fname.mode
            if 'b' in curr_mode:
                self.fObj = fname
            else:
                curr_mode += 'b'
                tempName = fname.name
                fname.close()
                self.fObj = open(tempName, curr_mode)
        else:
            raise IOError("fname must be a string or file")

        self.mode = self.fObj.mode

        self.password = password
        self.key = key
        self.bs = AES.block_size

        # Generate the key
        self.gen_key()

        # Setup the object's stream [used as a read buffer or write buffer]
        self.stream = ''
        self.fileOpen = True

    # Create the hashed password and initiation vector for the cipher
    def gen_key(self):

        # Support a pre-hashed key either saved in a file or passed as a string
        if isinstance(self.key, str):
            decoded_key = base64.b64decode(self.key)
            if self.key.endswith('.key'):
                self.key = open(self.key, 'rb').read()
            elif len(decoded_key) == 32:
                self.key = decoded_key
            else:
                msg = "Key {} not supported.".format(self.key)
                raise NotImplementedError(msg)
        # Also support a password hashed into a key
        elif isinstance(self.password, str):
            self.key = hashlib.sha256(self.password).digest()
        else:
            raise ValueError("Key or password is invalid.")

        # Have to create a new password from the key to use in IV
        # - allows for message recovery using key file without password
        self.password = base64.b64encode(self.key)

    def gen_iv(self):
        iv = ''
        d_i = ''
        while len(iv) < self.bs:
            d_i = hashlib.sha256(d_i + self.password + self.salt).digest()
            iv += d_i
        self.iv = iv[:self.bs]

    def check_mode(self, mType):
        if not self.mode.startswith(mType):
            msg = 'Requested operation not compatible with current file mode'
            raise IOError(msg)

    def close(self):
        self.fObj.close()

    # Allow use in a with statement
    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_value, traceback):
        self.close()

    # Add a weak method to ensure open encryption streams are closed properly
    def __del__(self):
        self.close()


crypto.py
from Crypto.Cipher import AES
from Crypto import Random
import hashlib
import base64


#  AES encrypted file-like object
#  - Support for Reader and Writer objects


class cryptoClass:
    def __init__(self, fname, password=None, key=None, mode=None):

        # Create a file handler
        if isinstance(fname, str):
            self.fObj = open(fname, mode)

        elif isinstance(fname, file):  # noqa: F821
            curr_mode = fname.mode
            if 'b' in curr_mode:
                self.fObj = fname
            else:
                curr_mode += 'b'
                tempName = fname.name
                fname.close()
                self.fObj = open(tempName, curr_mode)

        else:
            raise IOError

        self.mode = self.fObj.mode

        self.passwrd = password
        self.key = key
        self.bs = AES.block_size

        # Generate the key
        self.gen_key()

        # Setup the object's stream [used as a read buffer or write buffer]
        self.stream = ''
        self.fileOpen = True

    # Create the hashed password and initiation vector for the cipher
    def gen_key(self):

        # Support a pre-hashed key either saved in a file or passed as a string
        if isinstance(self.key, str) and self.key.endswith('.key'):
            self.key = open(self.key, 'rb').read()
        elif isinstance(self.key, str):
            if len(base64.b64decode(self.key)) == 32:
                self.key = base64.b64decode(self.key)

            if len(self.key) != 32:
                raise NotImplementedError

        # Also support a password hashed into a key
        elif isinstance(self.passwrd, str):
            self.key = hashlib.sha256(self.passwrd).digest()

        # Create a new key if the Writer object is being used
        should_create_new_key = (self.__class__.__name__ == 'Writer' and
                                 self.key is None and
                                 self.passwrd is None)
        if should_create_new_key:
            self.key = hashlib.sha256(Random.new().read(32)).digest()
            print('Encryption Key: {}\n'.format(base64.b64encode(self.key)))

        # Have to create a new password from the key to use in IV
        # - allows for message recovery using key file without password
        self.passwrd = base64.b64encode(self.key)

    def gen_iv(self):
        iv = ''
        d_i = ''
        while len(iv) < self.bs:
            d_i = hashlib.sha256(d_i + self.passwrd + self.salt).digest()
            iv += d_i
        self.iv = iv[:self.bs]

    def check_mode(self, mType):
        if not self.mode.startswith(mType):
            print('Requested operation not compatible with current file mode')
            raise IOError

    # Allow use in a with statement
    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_value, traceback):
        self.close()

    # Add a weak method to ensure open encryption streams are closed properly
    def __del__(self):
        self.close()


class Reader(cryptoClass):

    def __init__(self, fname, password=None, key=None, enableSeek=True):

        # Check for a key or password
        if key is None and password is None:
            print('Need either a password or a key (file) for decryption')
            raise NotImplementedError

        cryptoClass.__init__(self, fname, password, key, 'rb')

        # Verify the file object is in read mode
        self.check_mode('r')

        # Need to retrieve the salt from the file
        self.get_salt()
        self.gen_iv()

        # Initiate the decryptor
        self.cipher = AES.new(self.key, AES.MODE_CBC, self.iv)

        # Setup the object's stream [used as a read buffer or write buffer]
        self.streamLines = 0
        self.chunk_unprocessed = ''
        self.enableSeek = enableSeek

        # Allow for seek function; need to store previous values
        self.prev_stream = ''

    # Retrieve or Create the salt for the cipher
    def get_salt(self):
        self.salt = self.fObj.read(self.bs)[len('Salted__'):]

    # Method to read the next line
    # + decrypt enough to cover the next line
    def readline(self):
        while self.streamLines == 0 and self.fileOpen:
            self.decrypt_chunk()

        currLine, lineSuff, self.stream = self.stream.partition('\n')
        currChunk = currLine + lineSuff
        self.streamLines -= 1

        if self.enableSeek:
            self.prev_stream += currChunk

        return currChunk

    def read(self, size=None):
        while (len(self.stream) < size or size is None) and self.fileOpen:
            self.decrypt_chunk()
        if size is None:
            size = len(self.stream)
            currChunk, self.stream = self.stream, ''
        else:
            size = min(size, len(self.stream))
            currChunk = self.stream[:size]
            self.stream = self.stream[size:]

        if self.enableSeek:
            self.prev_stream += currChunk

        return currChunk

    def seek(self, pos, mode=0):
        if not self.enableSeek:
            print('seek disabled for this object')
            raise NotImplementedError

        # - Absolute position: from the start of the file
        # + equivalent to a relative position plus the
        #  length of the previous stream
        if mode == 0:
            pos -= len(self.prev_stream)

        # - Relative position: from the current position
        if mode == 0 or mode == 1:
            if pos < 0:
                self.stream = self.prev_stream[pos:] + self.stream
                self.prev_stream = self.prev_stream[:pos]
            else:
                self.read(pos)

        elif mode == 2:
            print('Seek from EOF is not implemented')
            raise IOError
        else:
            print('Unknown mode requested for seek')
            raise IOError

    def decrypt_chunk(self):
        self.chunk_processed = self.chunk_unprocessed
        crypted = self.fObj.read(1024 * self.bs)
        self.chunk_unprocessed = self.cipher.decrypt(crypted)
        if len(self.chunk_unprocessed) == 0:
            padding_length = ord(self.chunk_processed[-1])
            self.chunk_processed = self.chunk_processed[:-padding_length]
            self.fileOpen = False
        self.stream += self.chunk_processed
        self.streamLines += self.chunk_processed.count('\n')
        return 1

    def close(self):
        if not self.fObj.closed:
            self.fObj.close()


class Writer(cryptoClass):

    def __init__(self, fname, password=None, key=None, saveKey_file=None):
        cryptoClass.__init__(self, fname, password, key, 'wb')

        # Write the key to the specified file
        if saveKey_file is not None:
            open(saveKey_file, 'wb').write(self.key)

        # Verify the file object is in write mode
        self.check_mode('w')

        self.get_salt()
        self.gen_iv()

        # Initiate the encryptor
        self.cipher = AES.new(self.key, AES.MODE_CBC, self.iv)

    # Create the salt for the cipher
    def get_salt(self):
        # If it is a writable file need to
        # save the salt value to the start of the file
        self.salt = Random.new().read(self.bs - len('Salted__'))
        self.fObj.write('Salted__' + self.salt)

    def write(self, str):
        self.stream += str
        self.check_write_buffer()

    # Check if enough data is available to write an encrypted chunk
    def check_write_buffer(self):
        # - retain partial chunks and only pad if the file is being closed
        while len(self.stream) >= 1024 * self.bs:
            chunk = self.stream[:1024*self.bs]
            self.fObj.write(self.cipher.encrypt(chunk))
            self.stream = self.stream[1024*self.bs:]

    def close(self):
        if self.fObj.closed:
            return

        # Check for a partial string and pad if necessary
        if len(self.stream) == 0 or len(self.stream) % self.bs != 0:
            padding_length = (self.bs - len(self.stream) % self.bs) or self.bs
            self.stream += padding_length * chr(padding_length)
        else:
            self.fObj.write(self.cipher.encrypt(self.stream))
            self.stream = self.bs * chr(self.bs)

        self.fObj.write(self.cipher.encrypt(self.stream))
        self.fObj.close()
